### PR TITLE
feat: add support for dropping the database

### PR DESCRIPTION
**Pull request by alvincrespo**
This PR ensures that a developer can drop their database from the script `db:drop`

## Changes

- Adds the script `db:drop` to a generated apps `package.json`
- Includes Mocha for unit testing Bison
- Creates an `createApp` script to create the app but NOT start it (to verify setup)
- Defines a unit test to ensure that `DATABASE_NAME` is properly generated in `prisma/.env`
- Extends our `Release` workflow to include unit tests
- Creates a `test:prisma` script for independently testing prisma setup

## Checklist

- [x] Requires dependency update?
- [x] Generating a new app works

### CODE REVIEW

File: scripts/createAppAndStartServer.js [hunk 1]
```diff
@@ -1,19 +1,30 @@
...
+const fs = require("fs");
 const os = require("os");
 const path = require("path");
-const mkdtemp = promisify(require("fs").mkdtemp);
+const mkdtemp = promisify(fs.mkdtemp);
 const execa = require("execa");
 module.exports = {};
 
-async function createAppAndStartServer() {
+async function init() {
```
Comment (alvincrespo) [inline scripts/createAppAndStartServer.js:9]: Just keeping the main method calls in sync across scripts.

File: scripts/createAppAndStartServer.js [hunk 2]
```diff
@@ -1,19 +1,30 @@
...
   const appDir = await createApp(args);
   return startServer(appDir);
 }
 
+async function createTmpDir() {
+  const tmpDir = await mkdtemp(`${os.tmpdir()}${path.sep}`);
+
+  fs.writeFile("./tmp/tmpDir", tmpDir, () => {
+    console.log("tmpDir location stored.");
+  });
```
Comment (alvincrespo) [inline scripts/createAppAndStartServer.js:20]: We're storing the tmp directory location so we can pick it up easily if needed.

File: template/package.json.ejs [hunk 3]
```diff
@@ -11,6 +11,7 @@
     "cypress:local": "CYPRESS_LOCAL=true CYPRESS_BASE_URL=http://localhost:3000 cypress",
     "db:migrate": "yarn -s prisma migrate up --experimental",
     "db:setup": "yarn db:migrate && yarn prisma generate",
+    "db:drop": "source prisma/.env && dropdb $(echo $DATABASE_NAME)",
```
Comment (alvincrespo) [inline template/package.json.ejs:14]: `dropdb` is a command line helper for postgres. Comes with every install - more [here](https://www.postgresql.org/docs/9.3/app-dropdb.html).

**Comment by alvincrespo** [comment]
@cball I wasn't sure if we wanted to use Mocha for unit tests - but went down that route to at least get something running. Let me know if you have a library pref and I can update.

File: template/package.json.ejs (hunk 3 above)
Comment (cball) [reply template/package.json.ejs:14]: what do you think about putting this into `scripts/dropDatabase`? If we did that we could do:

```js
import { Client } from 'pg';

const connectionString = process.env.DATABASE_URL;
const db = new Client({ connectionString });
const { database } = db.connectionParameters;

async function drop() {
  await db.connect();
  console.log(`dropping the ${database} database`);
  await db.query(`DROP DATABASE IF EXISTS ${database}`);

  await db.end();
}

drop();
```

Comment (cball) [reply template/package.json.ejs:14]: the big advantage here I think is that we can just use the existing `DATABASE_URL` and not have to manage another variable. we should have `pg` included as a dependency already.

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -31,6 +31,9 @@ jobs:
       - name: Install packages
         run: |
           yarn install --frozen-lockfile
+      - name: Run Prisma Unit Tests
+        run: |
+          yarn test:prisma
```
Comment (cball) [inline .github/workflows/release.yml:36]: should we make this generic unit tests?

File: package.json [hunk 5]
```diff
@@ -44,6 +45,7 @@
     "@semantic-release/git": "^9.0.0",
     "cypress": "^5.1.0",
     "ejs-lint": "^1.1.0",
+    "mocha": "^8.1.3",
```
Comment (cball) [inline package.json:48]: Like I mentioned in the comment I think we should keep this as `jest` for continuity sake. glad you got this in though!

File: scripts/createApp.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+  await execa("node", [cliPath, name, ...cliOptions], {
+    cwd: tmpdir,
+    stdio: "inherit",
+  });
+
+  return path.join(tmpdir, name);
+}
+
+init();
```
Comment (cball) [inline scripts/createApp.js:47]: 👍

File: template/prisma/_.env.ejs [hunk 7]
```diff
@@ -5,3 +5,4 @@
 # See the documentation for all the connection string options: https://pris.ly/d/connection-strings
 
 DATABASE_URL="postgresql://<%= db.dev.user %><%= db.dev.password %>@<%= db.dev.host %>:<%= db.dev.port %>/<%= db.dev.name %>?schema=public"
+DATABASE_NAME="<%= db.dev.name %>"
```
Comment (cball) [inline template/prisma/_.env.ejs:8]: if we go the scripts route, may not need this

**Review by cball — CHANGES_REQUESTED** [review]
awesome @alvincrespo! It might be good to stick with `jest` since that's what we are leveraging once the app is generated. Looking at what you have for Mocha I don't think it will be much of an update.

Along the lines of what you mentioned, I'd be curious if our unit tests need to go through the whole app generation process, but we can circle back to that.

The biggest suggestion I have is around creating a script to drop the db. If we go that route, you'll want something like this:

```json
{
  "db:drop": "DOTENV_CONFIG_PATH=./prisma/.env ts-node -r dotenv/config ./scripts/dropdb",
}
```

**Comment by alvincrespo** [comment]
@cball Awesome recommendations! This makes sense to me. I'll go ahead and implement.

**Comment by alvincrespo** [comment]
@cball Ready for another round of review.

**Review by cball — APPROVED** [review]

**Pull request merged**